src/metabase/query_processor/middleware/cache_backend/db.clj

```diff
@@ -23,17 +28,71 @@
     :updated_at [:<= (du/->Timestamp (- (System/currentTimeMillis)
                                         (* 1000 (public-settings/query-caching-max-ttl))))]))
 
+(defn- throw-if-max-exceeded [max-num-bytes bytes-in-flight]
+  (when (< max-num-bytes bytes-in-flight)
+    (throw (ex-info "Exceeded the max number of bytes" {:type ::max-bytes}))))
+
+(defn- limited-byte-output-stream
+  "Returns a `FilterOutputStream` that will throw an exception if more than `max-num-bytes` are written to
+  `output-stream`"
+  [max-num-bytes output-stream]
+  (let [bytes-so-far (atom 0)]
+    (proxy [java.io.FilterOutputStream] [output-stream]
+      (write
+        ([byte-or-byte-array]
+         (let [^java.io.OutputStream this this]
+           (if-let [^bytes byte-arr (and (bytes? byte-or-byte-array)
+                                         byte-or-byte-array)]
+             (do
+               (swap! bytes-so-far + (alength byte-arr))
+               (throw-if-max-exceeded max-num-bytes @bytes-so-far)
+               (proxy-super write byte-arr))
+
+             (let [^byte b byte-or-byte-array]
+               (swap! bytes-so-far inc)
+               (throw-if-max-exceeded max-num-bytes @bytes-so-far)
+               (proxy-super write b)))))
+        ([byte-arr offset length]
+         (let [^java.io.OutputStream this this]
+           (swap! bytes-so-far + length)
+           (throw-if-max-exceeded max-num-bytes @bytes-so-far)
+           (proxy-super write byte-arr offset length)))))))
+
+(defn- compress-until-max
+  "Compresses `results` and returns a byte array. If more than `max-bytes` is written, `::exceeded-max-bytes` is
+  returned."
+  [max-bytes results]
+  (try
+    (let [bos  (ByteArrayOutputStream.)
+          lbos (limited-byte-output-stream max-bytes bos)]
+      (with-open [buff-out (BufferedOutputStream. lbos)
+                  gz-out   (GZIPOutputStream. buff-out)
+                  data-out (DataOutputStream. gz-out)]
+        (nippy/freeze-to-out! data-out results))
```

Comment:
The previous code took care to not fully realize the `lazy-seq` when serializing the rows to a string. Calling `freeze-to-out!` does realize the `lazy-seq`. With the code as it is today, there are other things that are already fully realizing the seq and holding onto it. So this isn't great as it's another place that it's happening, I did test out that streaming would be supported. If we altered our results format to one where the results could be a `doseq` with calls to `freeze-to-out!` it would support not fully realizing the `lazy-seq`.

Reply:
what are the things fully realizing the seq?

Reply:
The previous code fully realized the seq for anything less than the max cache setting. It did that both in the serialization code that checked the size and the code to compress/serialize it.

I think there were several places in the query processor that would fully realize it before the refactor. After it looks like [this](https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/add_row_count_and_status.clj#L16) will fully realize it along with the caching code. Also due to the fact that the lazy-seq of rows is a value in a map, the map is holding onto the head of the lazy seq. With that, serializing the rows to JSON would also fully realize the seq, and hold it in memory.

Reply:
Oh yeah. Maybe we can remove that? Seems like for smaller results sets we can do the count on the frontend and for large ones they'll be output to CSV or something else that doesn't include the count anyway.

Alternatively we could create some sort of implementation of count that counted the rows as they were written to the output stream and then add the count at the end of the response after the results finished streaming